docs/man/canedalib.1
.TH CanedaLib "1" "July 2005" "Debian/GNU Linux" "User Commands"
.SH NAME
CanedaLib \- A component library manager for Caneda.
.SH SYNOPSIS
.B canedalib
[\fIOPTION\fR] ...
.SH DESCRIPTION

\fBCaneda\fR is an integrated circuit simulator which means you are able
to setup a circuit with a graphical user interface (GUI) and simulate
the large-signal, small-signal and noise behaviour of the circuit.
After that simulation has finished you can view the simulation results
on a presentation page or window.

The software aims to support all kinds of circuit simulation types,
e.g. DC, AC, S-parameter, harmonic balance analysis, noise analysis,
etc.

\fBCanedaLib\fR is a program to manage Caneda component libraries. On the
left side of the application window the available libraries can be
browsed to find the wanted component.  By clicking on the component
name its description can be seen on the right side. The selected
component is transported to the Caneda application by clicking on the
button "Copy to Clipboard".  Being back in the schematic window the
component can be inserted by pressing CTRL-V (paste from clipboard).

.SH AVAILABILITY
The latest version of Caneda can always be obtained from
\fBwww.caneda.tuxfamily.org\fR
.SH "REPORTING BUGS"
Known bugs are documented within the BUGS file.  Report bugs to
<caneda-bugs@lists.tuxfamily.org>.
.SH COPYRIGHT
Copyright \(co 2005 Michael Margraf <michael.margraf@alumni.tu-berlin.de>
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
Written by Michael Margraf <michael.margraf@alumni.tu-berlin.de> and
Stefan Jahn <stefan@lkcc.org>.


ChangeLog
2010-03-25 Renamed project to Caneda
	* Ported all clases and sources to new project name.

2007-10-10 Bastien Roucaries initiated api documentation.
	* qucs/node.h, qucs/node.cpp, qucs/wire.h, qucs/wire.cpp,
	qucs/wireline.h, qucs/item.h, qucs/item.cpp: Add doxygen comments.

2007-03-05  Martín Ribelotta and Lisandro Damián Nicanor Pérez Meyer
	* Ported qucs-edit to qt4
	* Ported qucs-lib to qt4

2006-09-11  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Don't link using "-s" on MacOSX since it drops lazy
	dynamic symbol information.  Also checking for {moc|uic|lrelease|
	lupdate}-qt4 programs.  Added some more paths for Qt headers and
	libraries.

2006-08-08  Stefan Jahn  <stefan@lkcc.org>

	* qucs-attenuator/qucsattenuator.cpp (slotSetText_Zin):
	Synchronized Zout and Zin when specifying a bridged-tee.

	* qucs-attenuator/attenuatorfunc.cpp (Calc): Fixed computation
	routine for bridged-tee topology.

2006-05-29  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.9 and bumped up to version
	0.0.10.

2006-05-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Added -s (strip) to linker flags when no
	debugging enabled.
	Put MinGW32 checks in place, thus it is now possible to cross-
	compile the qucs package using a GNU/Linux box.  Had to separate
	LDFLAGS and LIBS because linker command line order matters (-lqt-mt
	-lqtmain *must* be at the end of the command line), otherwise
	linking fails.  Changed rules in appropriate Makefile.am's.
	Additionally checking for the 'ar' tool.

2006-02-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-transcalc/main.cpp, qucs-lib/main.cpp,
	qucs-help/main.cpp, qucs-filter/main.cpp, qucs-edit/main.cpp
	(loadSettings): Try loading items from main qucsrc even if
	application specific resource file does not exist.

2006-02-10  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac (paths): Added $QTDIR/lib64 to the library path
	to look Qt libraries for.  For 64-bit GNU/Linux's.

2006-01-25  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_cauer.cpp: Fixed band stop filters.  Now
	correctly working and exported into the clipboard.

2006-01-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.8 and bumped up to version
	0.0.9.

2006-01-18  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Lowpass/Highpass/Bandpass of
	Cauer filter now quite like working.

	* qucs-filter/qf_cauer.cpp: Fixed bug in bandpass/bandstop
	generation.

2006-01-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Preliminary support for Cauer
	filters.  Many bugs here...

2006-01-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Checking for Qt programs also in the
	$QTDIR/bin path.

2006-01-07  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qucsfilter.cpp (calculateFilter): Built in
	the cauer filter code by Vincent Habchi.

2005-11-15  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* show cursor position in qucs-edit
	* extend Bessel coefficient table to order 19

2005-11-08  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* fixed small bug in library search

2005-10-18  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Fixed overwriting CXXFLAGS and CFLAGS
	environment variables via 'configure' command line.

2005-10-13  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new libraries: Bridges, LEDs

2005-08-19  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new components models to library, thanks to
	Leandro D'Archivio

2005-07-23  Stefan Jahn  <stefan@lkcc.org>

	* qucs-lib/qucslib.1: Added manpage for the QucsLib library
	manager program.

	* configure.ac: Finally fixed freebsd[456]* detection.

2005-07-22  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.7 and bumped up to version
	0.0.8.

2005-05-25  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.6 and bumped up to version
	0.0.7.

2005-05-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs*.1: Updated manpages.

2005-03-01  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.5 and bumped up to version
	0.0.6.

2005-02-27  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp, qucs-edit/main.cpp (main): Made the
	QucsHelp and QucsEdit applications relocatable using the
	environment variable QUCSDIR.

2005-02-23  Stefan Jahn  <stefan@lkcc.org>

	* THANKS: Corrected Toyoyuki Ishikawa's name at all locations
	in the project.

2005-02-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: LDFLAGS additions for FreeBSD users.

2005-02-18  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* configure.ac: Add another paths to look for Qt header/library in.
	Required for FreeBSD. Use gmake there.

2004-11-24  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* autogen.sh: Make aclocal and automake program be overridable
	with environment variable. Need this so i can easily build on
	debian woody, I hope nobody takes offense.
	PS: While Stefan is away, I will commit his stuff.

2004-11-23  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* added icon to QucsEdit and QucsHelp

2004-11-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.cpp (slotAbout): Added proper copyright
	notice to the 'About' dialog.

2004-10-23  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.4 and bumped up to version
	0.0.5.

2004-09-30  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.1: Added simple man page for qucsedit.

	* qucs-help/qucshelp.1: Added simple man page for qucshelp.

	* qucs-help/main.cpp (main): Trying to obtain the correct
	language help directory by splitting e.g. 'es_MX' to use the
	'es' directory.

2004-09-15  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp (main): Using locale() string to check
	whether the documentation directory in the specific language
	is available.

	* configure.ac: Made the DOCDIR variable for choosing a
	specific language.

2004-09-04  Stefan Jahn  <stefan@lkcc.org>


caneda-lib/canedalib.cpp
/***************************************************************************
                               canedalib.cpp
                              -------------
    begin                : Sat May 28 2005
    copyright            : (C) 2005 by Michael Margraf
    email                : michael.margraf@alumni.tu-berlin.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <QtGui/QMenu>
#include <QtGui/QMenuBar>
#include <QtGui/QAction>
#include <QtGui/QComboBox>
#include <QtGui/QClipboard>
#include <QtGui/QApplication>
#include <QtGui/QVBoxLayout>
#include <QtGui/QHBoxLayout>
#include <QtGui/QGroupBox>
#include <QtGui/QPushButton>
#include <QtGui/QMessageBox>
#include <QtGui/QTextEdit>
#include <QtGui/QListWidget>
#include <QtCore/QRegExp>
#include <QtGui/QCloseEvent>
#include <QtCore/QTextStream>

#include "canedalib.h"
#include "librarydialog.h"
#include "displaydialog.h"
#include "symbolwidget.h"
#include "searchdialog.h"

/* Constructor setups the GUI. */
CanedaLib::CanedaLib()
{
  // set application icon
  setWindowIcon (QPixmap(CanedaSettings.BitmapDir + "big.caneda.xpm"));
  setWindowTitle("Caneda Library Tool " PACKAGE_VERSION);

  QMenuBar * menuBar = new QMenuBar (this);

  // create file menu
  QMenu * fileMenu = menuBar->addMenu(tr("&File"));//new QMenu ();
  QAction * manageLib = new QAction (tr("Manage User &Libraries..."), this);
  manageLib->setShortcut(QKeySequence("CTRL+M"));

  fileMenu->addAction ( manageLib );
  connect(manageLib, SIGNAL(activated()), SLOT(slotManageLib()));

  fileMenu->addSeparator();

  QAction * fileQuit = new QAction (tr("&Quit"), this);
  fileQuit->setShortcut(QKeySequence("CTRL+Q"));

  fileMenu->addAction(fileQuit);
  connect(fileQuit, SIGNAL(activated()), SLOT(slotQuit()));

  // Insert separator
  menuBar->addSeparator ();

  // create help menu
  QMenu * helpMenu = menuBar->addMenu(tr("&Help"));
  QAction * helpHelp = new QAction (tr("&Help"), this);
  helpHelp->setShortcut(QKeySequence("F1"));

  helpMenu->addAction (helpHelp);
  connect(helpHelp, SIGNAL(activated()), SLOT(slotHelp()));
  QAction * helpAbout = new QAction(tr("About"),helpMenu);

  helpMenu->addAction (helpAbout);
  connect(helpAbout, SIGNAL(activated()), SLOT(slotAbout()));

  // main box
  QVBoxLayout * all = new QVBoxLayout (this);
  all->setSpacing (0);
  all->setMargin (0);

  // reserve space for menubar
  QWidget * Space = new QWidget (this);
  Space->setFixedSize(5, menuBar->height() + 2);
  all->addWidget (Space);

  // main layout
  QHBoxLayout * h = new QHBoxLayout();
  h->setSpacing (5);
  h->setMargin (3);
  all->addLayout (h);

  // library and component choice
  QGroupBox * LibGroup = new QGroupBox (tr("Component Selection"),this);
  h->addWidget(LibGroup);
  QVBoxLayout *grpl = new QVBoxLayout( LibGroup );
  Library = new QComboBox (LibGroup);
  grpl->addWidget( Library );
  connect(Library, SIGNAL(activated(int)), SLOT(slotSelectLibrary(int)));
  CompList = new QListWidget(LibGroup);
  grpl->addWidget( CompList );
//  connect(CompList, SIGNAL(highlighted(QListBoxItem*)),SLOT(slotShowComponent(QListBoxItem*)));
  connect(CompList, SIGNAL(currentItemChanged(QListWidgetItem*,QListWidgetItem*)),
	  this, SLOT(slotShowComponent(QListWidgetItem*,QListWidgetItem*)));

  QHBoxLayout * h1 = new QHBoxLayout();
  grpl->addLayout( h1 );
  QPushButton * SearchButton = new QPushButton (tr("Search..."), LibGroup);
  h1->addWidget( SearchButton );
  connect(SearchButton, SIGNAL(clicked()), SLOT(slotSearchComponent()));
  QWidget * w1 = new QWidget(LibGroup);
  h1->addWidget(w1);
  h1->setStretchFactor(w1, 5); // stretchable placeholder

  // component display
  QGroupBox *CompGroup = new QGroupBox (tr("Component"), this);
  h->addWidget( CompGroup );
  QVBoxLayout * compLayout = new QVBoxLayout(CompGroup);
  CompDescr = new QTextEdit(CompGroup);
  compLayout->addWidget(CompDescr);
  CompDescr->setReadOnly(true);
  CompDescr->setWordWrapMode(QTextOption::NoWrap);

  Symbol = new SymbolWidget (CompGroup);
  compLayout->addWidget( Symbol );

  QHBoxLayout * h2 = new QHBoxLayout();
  compLayout->addLayout( h2  );
  QPushButton * CopyButton = new QPushButton (tr("Copy to clipboard"), CompGroup);
  h2->addWidget( CopyButton );
  connect(CopyButton, SIGNAL(clicked()), SLOT(slotCopyToClipBoard()));
  QPushButton * ShowButton = new QPushButton (tr("Show Model"), CompGroup);
  h2->addWidget( ShowButton );
  connect(ShowButton, SIGNAL(clicked()), SLOT(slotShowModel()));

  // ......................................................
  putLibrariesIntoCombobox();
}

/* Destructor destroys the application. */
CanedaLib::~CanedaLib()
{
}

// ----------------------------------------------------
// Put all available libraries into ComboBox.
void CanedaLib::putLibrariesIntoCombobox()
{
  Library->clear();

  UserLibCount = 0;
  QStringList LibFiles;
  QStringList::iterator it;
  if(UserLibDir.cd(".")) { // user library directory exists ?
    LibFiles = UserLibDir.entryList(QStringList("*.lib"), QDir::Files, QDir::Name);
  UserLibCount = LibFiles.count();

    for(it = LibFiles.begin(); it != LibFiles.end(); it++)
      Library->addItem((*it).left((*it).length()-4));
  }

  QDir LibDir(CanedaSettings.LibDir);
  LibFiles = LibDir.entryList(QStringList("*.lib"), QDir::Files, QDir::Name);

  for(it = LibFiles.begin(); it != LibFiles.end(); it++)
    Library->addItem((*it).left((*it).length()-4));

  slotSelectLibrary(0);
}

// ----------------------------------------------------
void CanedaLib::slotAbout()
{
  QMessageBox::about(this, tr("About..."),
    "CanedaLib Version " PACKAGE_VERSION "\n"+
    tr("Library Manager for Caneda\n")+
    tr("Copyright (C) 2005 by Michael Margraf\n")+
    "\nThis is free software; see the source for copying conditions."
    "\nThere is NO warranty; not even for MERCHANTABILITY or "
    "\nFITNESS FOR A PARTICULAR PURPOSE.");
}

// ----------------------------------------------------
void CanedaLib::slotQuit()
{
  int tmp;
  tmp = x();		// call size and position function in order to ...
  tmp = y();		// ... set them correctly before closing the ...
  tmp = width();	// dialog !!!  Otherwise the frame of the window ...
  tmp = height();	// will not be recognized (a X11 problem).

  accept();
}

// ----------------------------------------------------
void CanedaLib::closeEvent(QCloseEvent *Event)
{
  int tmp;
  tmp = x();		// call size and position function in order to ...
  tmp = y();		// ... set them correctly before closing the ...
  tmp = width();	// dialog !!!  Otherwise the frame of the window ...
  tmp = height();	// will not be recognized (a X11 problem).

  Event->accept();
}

// ----------------------------------------------------
void CanedaLib::slotManageLib()
{
  (new LibraryDialog(this))->exec();
  putLibrariesIntoCombobox();
}

// ----------------------------------------------------
void CanedaLib::slotHelp()
{
  DisplayDialog *d = new DisplayDialog(this);
  d->setWindowTitle("CanedaLib Help");
  d->resize(250, 325);
  d->Text->setText(
     tr("CanedaLib is a program to manage Caneda component libraries. "
	"On the left side of the application window the available "
	"libraries can be browsed to search the wanted component. "
	"By clicking on the component name its description can be "
	"seen on the right side. The selected component is "
        "transported to the Caneda application by clicking on the "
	"button \"Copy to Clipboard\". Being back in the schematic "
	"window the component can be inserted by pressing CTRL-V "
	" (paste from clipboard).") + "\n" +
     tr("A more comfortable way: The component can also be placed "
        "onto the schematic by using Drag n'Drop."));
  d->show();
}

// ----------------------------------------------------
void CanedaLib::slotCopyToClipBoard()
{
  QString s = "<Caneda Schematic " PACKAGE_VERSION ">\n";
  s += "<Components>\n  " +
       Symbol->theModel() +
       "\n</Components>\n";

  // put resulting schematic into clipboard
  QClipboard *cb = QApplication::clipboard();
  cb->setText(s);
}

// ----------------------------------------------------
void CanedaLib::slotShowModel()
{
  DisplayDialog *d = new DisplayDialog(this);
  d->setWindowTitle("Model");
  d->resize(500, 150);
  d->Text->setText(Symbol->ModelString);
  d->Text->setWordWrapMode(QTextOption::NoWrap);
  d->show();
}

// ----------------------------------------------------
void CanedaLib::slotSelectLibrary(int Index)
{
  int Start, End, NameStart, NameEnd;
  End = Library->count()-1;
  if(Library->itemText(End) == tr("Search result"))
    if(Index < End)
      Library->removeItem(End); // if search result still there -> remove it
    else  return;


  CompList->clear();
  LibraryComps.clear();
  DefaultSymbol = "";

  QFile file;
  if(Index < UserLibCount)  // Is it user library ?
    file.setFileName(UserLibDir.absolutePath() + QDir::separator() + Library->itemText(Index) + ".lib");
  else
    file.setFileName(CanedaSettings.LibDir + Library->itemText(Index) + ".lib");

  if(!file.open(QIODevice::ReadOnly))
  {
    QMessageBox::critical(this, tr("Error"),
        tr("Cannot open \"%1\".").arg(
           CanedaSettings.LibDir + Library->itemText(Index) + ".lib"));
    return;
  }

  QTextStream ReadWhole(&file);
  QString LibraryString = ReadWhole.readAll();
  file.close();

  Start = LibraryString.indexOf("<Caneda Library ");
  if(Start < 0) {
    QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
    return;
  }
  End = LibraryString.indexOf('>', Start);
  if(End < 0) {
    QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
    return;
  }
  QString LibName = LibraryString.mid(Start, End-Start).section('"', 1, 1);

  Start = LibraryString.indexOf("\n<", End);
  if(Start < 0) return;
  if(LibraryString.mid(Start+2, 14) == "DefaultSymbol>") {
    End = LibraryString.indexOf("\n</DefaultSymbol>");
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }

    DefaultSymbol = LibraryString.mid(Start+16, End-Start-16);
    Start = End + 3;
  }

  while((Start=LibraryString.indexOf("\n<Component ", Start)) > 0) {
    Start++;
    NameStart = Start + 11;
    NameEnd = LibraryString.indexOf('>', NameStart);
    if(NameEnd < 0)  continue;

    End = LibraryString.indexOf("\n</Component>", NameEnd);
    if(End < 0)  continue;
    End += 13;

    CompList->addItem(LibraryString.mid(NameStart, NameEnd-NameStart));
    LibraryComps.append(LibName+'\n'+LibraryString.mid(Start, End-Start));
    Start = End;
  }

  ///FIXME (Is this alright?)
  CompList->setCurrentRow(0);  // select first item
}

// ----------------------------------------------------
void CanedaLib::slotSearchComponent()
{
  SearchDialog *d = new SearchDialog(this);
  if(d->exec() == QDialog::Accepted)
    QMessageBox::information(this, tr("Result"),
                             tr("No appropriate component found."));
}

// ----------------------------------------------------
void CanedaLib::slotShowComponent(QListWidgetItem *Item, QListWidgetItem*)
{
  if(!Item) return;

  QString CompString = LibraryComps.at(CompList->row(Item));
  QString LibName = CompString.section('\n', 0, 0);
  CompDescr->setText("Name: " + Item->text());
  CompDescr->append("Library: " + LibName);
  CompDescr->append("----------------------------");

  if(Library->currentIndex() < UserLibCount)
    LibName = UserLibDir.absolutePath() + QDir::separator() + LibName;


  int Start, End;
  Start = CompString.indexOf("<Description>");
  if(Start > 0) {
    Start += 13;
    End = CompString.indexOf("</Description>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    CompDescr->append(
      CompString.mid(Start, End-Start).replace(QRegExp("\\n\\x20+"), "\n").remove(0, 1));
  }


  Start = CompString.indexOf("<Model>");
  if(Start > 0) {
    Start += 7;
    End = CompString.indexOf("</Model>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    Symbol->ModelString =
      CompString.mid(Start, End-Start).replace(QRegExp("\\n\\x20+"), "\n").remove(0, 1);

    if((Symbol->ModelString.count('\n')) < 2)
      Symbol->createSymbol(LibName, Item->text());
  }


  Start = CompString.indexOf("<Symbol>");
  if(Start > 0) {
    Start += 8;
    End = CompString.indexOf("</Symbol>", Start);
    if(End < 0) {
      QMessageBox::critical(this, tr("Error"), tr("Library is corrupt."));
      return;
    }
    Symbol->setSymbol(CompString.mid(Start, End-Start),
                      LibName, Item->text());
  }
  else if(!DefaultSymbol.isEmpty())   // has library a default symbol ?
    Symbol->setSymbol(DefaultSymbol, LibName, Item->text());
}
